client/blocks/payment-request/hooks.js

```diff
@@ -0,0 +1,356 @@
+		 */
+		const performPayment = ( evt ) => ( createOrderResponse ) => {
+			if ( createOrderResponse.result === 'success' ) {
+				evt.complete( 'success' );
+
+				const partials = getRedirectUrlPartials(
+					createOrderResponse.redirect
+				);
+
+				// If no information is embedded in the URL that means the payment doesn't need
+				// verification and we can proceed as usual.
+				if ( ! partials || partials.length < 4 ) {
+					window.location = createOrderResponse.redirect;
+					return;
+				}
+
+				const { type, clientSecret, redirectUrl } = partials;
+
+				// The payment requires 3DS verification, so we try to take care of that here.
+				requestIntentConfirmation( type, clientSecret )
+					.then( handleIntentConfirmation( redirectUrl, type ) )
+					.catch( ( error ) => {
+						setExpressPaymentError( error.message );
+
+						// Report back to the server.
+						$.get( redirectUrl + '&is_ajax' );
+					} );
+			} else {
+				evt.complete( 'fail' );
+				setExpressPaymentError( createOrderResponse.messages );
+			}
+		};
+
+		const paymentMethodUpdateHandler = paymentRequest?.on(
+			'source',
+			( evt ) => {
+				const allowPrepaidCards =
+					wc_stripe_payment_request_params?.stripe
+						?.allow_prepaid_card === 'yes';
+
+				// Check if we allow prepaid cards.
+				if (
+					! allowPrepaidCards &&
+					evt?.source?.card?.funding === 'prepaid'
+				) {
+					setExpressPaymentError(
+						__(
+							"Sorry, we're not accepting prepaid cards at this time.",
+							'woocommerce-gateway-stripe'
+						)
+					);
+				} else {
+					// Create the order and attempt to pay.
+					createOrder( evt, paymentRequestType ).then(
+						performPayment( evt )
+					);
+				}
+			}
+		);
+
+		return () => {
+			paymentMethodUpdateHandler?.removeAllListeners();
+		};
+	}, [ stripe, paymentRequest, paymentRequestType, setExpressPaymentError ] );
+};
+
+/**
+ * Returns an onClick handler for payment request buttons. Resets the error state, syncs the
+ * payment request with the block, and calls the provided click handler.
+ *
+ * @param {Object}   paymentRequest - The Payment Request object.
+ * @param {Function} setExpressPaymentError - Used to set the error state.
+ * @param {Function} onClick - The onClick function that should be called on click.
+ * @param {Object}   billing - The billing data from the checkout or cart block.
+ *
+ * @return {Function} An onClick handler for the payment request buttons.
+ */
+export const useOnClickHandler = (
+	paymentRequest,
+	setExpressPaymentError,
+	onClick,
+	billing
+) => {
+	return useCallback( () => {
+		// Reset any Payment Request errors.
+		setExpressPaymentError( '' );
+
+		// Update the payment request with new billing information.
+		if ( paymentRequest ) {
+			updatePaymentRequest( {
+				paymentRequest,
+				total: billing.cartTotal,
+				currencyCode: billing.currency.code.toLowerCase(),
+				cartTotalItems: billing.cartTotalItems,
+			} );
```

Comment:
`usePaymentRequest` is using `getCartDetails()` to retrieve totals and display items.

There's this inconsistency for the labels when retrieving from the `billing` prop or from the AJAX request.

I don't have an idea on how to fix this yet, but perhaps we can still use the `billing` and `shippingData` prop in `usePaymentRequest` so we save one unnecessary request here.

Reply:
Hmmm, I'd like the labels to be the same for shortcode and Blocks 🤔 
That's why _ideally_ we'd use the AJAX API to set the label, but you're right; when we populate with information from the block that changes the label.

I'll look into this. Nice catch!

Reply:
Hmm, I'm honestly not sure how to fix this. I think we have to do 1 of 3 things:

1. Use `getCartDetails()` for everything in blocks and spend the time to make sure the shortcode and blocks PRBs populate the payment request with the same data
1. Use `getCartDetails()` for everything in blocks and accept the inconsistencies.
1. Use the AJAX API for everything (both blocks and shortcode) and accept that we have to make a request for the cart each time we call `updatePaymentRequest`.

The downside of (1) is extra maintenance — changes in one place will need changes in another place and those are really difficult to remember.

The downside of (2) is the inconsistency — The payment requests dialogs won't have the exact same information. When we start customizing them (e.g. when merchants enable the hook that adds an itemized list of things being purchased) we'll have to implement that twice, possibly in 2 different ways.

The downside of (3) is extra network requests.

I think (3) is the best option. It's a small network request we're trading for what would be a significant maintenance burden.

Reply:
> and `getCartDetails()` is making the extra request

ugh, sorry, I meant "use the Blocks data (i.e. front-end data) to populate the payment request" instead of `getCartDetails()` 🤦 

To clear this up a bit more: The options I meant to propose:

1. Use the [front-end data](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2159ef3b5697dc7265ea982161dbc9a90fc009bc/client/blocks/payment-request/use-initialization.js#L70-L77) to populate the payment request and fix up the labels so Shortcode and Blocks Payment Requests look the same.
2. Use the [front-end data](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/2159ef3b5697dc7265ea982161dbc9a90fc009bc/client/blocks/payment-request/use-initialization.js#L70-L77) to populate the payment request and accept any inconsistencies Shortcode and Blocks Payment Requests.
3. Use `getCartDetails()` (the AJAX API) to populate both Shortcode and Blocks Payment Requests.

The downsides list I added above is based on this ^ list, not the one in my original comment.

Sorry for the misunderstanding!

> But I agree, if we use getCartDetails() for everything, including updatePaymentRequest, we will have more consistency with the labels between blocks and shortcode, but the downside is the extra request.
>
> I think it's fine to use it in both places.

Just to be clear; you mean it's fine to use `getCartDetails()` in both places, yes? 🙂

Reply:
Ok, so I just made the change to use `getCartDetails()` for both creating the payment request, as well as updating it in 17af3684fe75db46fab8aa3b8d29a8909692f78a. I had to change the `onClick` handler a bit to accomodate the network request, but I honestly think it makes more sense the way it is now ☺️

Reply:
🤔 I was wondering just now - can the currency code change from the cart block? I've seen the previous implementation use `billing.currency.code.toLowerCase()`.

If so, we'd need to add `billing.currency.code` to the above suggestion, and somehow return the cart currency code from `getCartDetails()`.

Reply:
> I think a better approach would be to reset the button only when needsShipping or billing changes.

Yeah, that makes sense.

Fixed in 8924d4c16e12c97b9d9b23fcea5e24d189067948.

> There's a warning Do not call show() ... when you click the button.

To be fair, that happens even though I'm calling the `preventDefault()` function provided by Stripe 🙂